ISPN-14082: Document cache naming character limits

dfitzmau opened the pull request:
JIRA: https://issues.redhat.com/browse/ISPN-14082

**Needs backport to 13.0.**

dfitzmau commented:
Backport label applied, but updates for 13.0 require less content.

dfitzmau commented:
Hi @jabolina .

I have written a draft note or notes for the cache name/file name character limit issue.

I've lumped two notes into one for the purposes of a preliminary review.

I'm suggesting to add the cache name restriction to various cache type sectons outlined in the Infinispan caches guide, such as [Local caches](https://infinispan.org/docs/stable/titles/configuring/configuring.html#local-caches_caches).

I'm suggesting to add the file name restriction to a file-specific section in the Infinispan caches guide, such as [File-based cache stores](https://infinispan.org/docs/stable/titles/configuring/configuring.html#file-stores_persistence).

WDYT?

jabolina commented on documentation/src/main/asciidoc/topics/con_local_caches.adoc:
```diff
@@ -8,6 +8,18 @@ write-behind to persistent storage as well as management capabilities such as ev
+[IMPORTANT]
+====
+{brandname} places a restriction of less than or equal to 255 characters on a cache name or a cache template name.
+This restriction reduces server failure issues that can arise with long cache names or cache template names.
```
The restriction was because an internal limitation (in the byte representation we use during the cache creation IIRC) that will always fail with names longer than 255 bytes. We could rephrase this or remove.

dfitzmau replied:
Hey @jabolina . Thanks for reviewing the content and for adding feedback. 

My meaning behind the restriction was more focused on why this restriction has come into play. Might I ask what was the previous behavior? Why a character limit of 255 is required. What happened when a user created a cache and named it with + 255 characters?

jabolina replied:
The limitation already existed due to the way we represented the name during the creation. 

https://github.com/infinispan/infinispan/blob/22d7148c5b63d262f7bd0bb33d0c4f97de38a3a6/core/src/main/java/org/infinispan/util/ByteString.java#L36-L38

Originally an unchecked exception was thrown, which could crash the server, with the change we kind of formalized this limit, now we throw an exception with a proper feedback and explanation.

jabolina commented:
Only a minor thing about the text, but looking good. Next week I'll take a look about the location we can add these. If we have a doc dedicated to configuration would be the best place I think.

dfitzmau commented:
> 

Thanks!

Yeah I was thinking the [Configuring caches](https://infinispan.org/docs/stable/titles/configuring/configuring.html) document. Notes might need to go into a few sections as different cache types and templates exist in the doc. 

Maybe I could add a note at the beginning of a document chapter, such as [Infinispan caches](https://infinispan.org/docs/stable/titles/configuring/configuring.html#caches).

jabolina commented:
> 

Nice, I think that is the place for us to add this. Adding a waning note in the beginning of section 3..

jabolina commented on documentation/src/main/asciidoc/stories/assembly_cache_configuration.adoc:
```diff
@@ -10,6 +10,19 @@ For instance, you can configure {brandname} clusters to use replicated caches or
+[IMPORTANT]
+====
+{brandname} places a restriction of less than or equal to 255 characters on a cache name or a cache template name.
+If you exceed this character limit, {brandname} throws an exception message
```
Just a really small thing, the same in the other file.

```suggestion
If you exceed this character limit, {brandname} throws an exception.
```

dfitzmau replied:
Updated.

jabolina reviewed (commented):
Just a small change. And two things, I think it would be nice to have to something saying that shorter names are preferable, something in those lines. And we can add this to only one of the files to avoid duplicating, whichever one you prefer is good to me.

dfitzmau commented:
> Just a small change. And two things, I think it would be nice to have to something saying that shorter names are preferable, something in those lines. And we can add this to only one of the files to avoid duplicating, whichever one you prefer is good to me.

Thanks, @jabolina . I added another sentence to each admonition to state to keep names short. Please let me know your thoughts.

jabolina commented on documentation/src/main/asciidoc/stories/assembly_cache_configuration.adoc:
```diff
@@ -10,6 +10,25 @@ For instance, you can configure {brandname} clusters to use replicated caches or
+
+[IMPORTANT]
+====
+A local host's file system might place a character limitation on a file name.
+Check the file system's configuration for any specified character limitations, and ensure that your file name does not exceed this limitation. 
+
+If your file name exceeds a file system's character limitation, the write operation for adding a file name to the file system might fail.
+
+Try to write succinct file names.
```
We should rephrase this a little. Something like:

The file system may impose limitations on the length of a file name, so ensure that a cache name does not exceed this limitation. If the cache name exceeds the file system limitations, operations and/or initialization might fail.

dfitzmau replied:
Thanks, @jabolina .

I updated the content based on your feedback. 

Would it be fair to state that initialization is a type of operation? 

I stated "general operations or initialing operations towards that cache might fail", because numerous entities might perform operations on a cache. Is this accurate?

jabolina replied:
I think it should be OK.

jabolina reviewed (approved):
LGTM. @dvagnero if you could take a look too, please.

dvagnero commented:
Hi, sorry for the delay. I think this content fits in the documentation/src/main/asciidoc/stories/assembly_cache_configuration.adoc (chapter 3). I think that's where users will look for such information.
The documentation/src/main/asciidoc/stories/assembly_cache_modes.adoc (chapter 1) summarizes use cases.  
These two assemblies are both in the same guide and aren't shared in other guides so I'd suggest removing it from chapter 1. WDYT @dfitzmau ?

dfitzmau commented:
> Hi, sorry for the delay. I think this content fits in the documentation/src/main/asciidoc/stories/assembly_cache_configuration.adoc (chapter 3). I think that's where users will look for such information. The documentation/src/main/asciidoc/stories/assembly_cache_modes.adoc (chapter 1) summarizes use cases. These two assemblies are both in the same guide and aren't shared in other guides so I'd suggest removing it from chapter 1. WDYT @dfitzmau ?

HI @dvagnero . I wanted to put the notes in Chapter 1, so the notes display earlier in the guide to grab the user's attention. 

I have removed the notes from chapter ; they now only display in Chapter 3. Please let me know if this is good.

dfitzmau commented:
> > Hi, sorry for the delay. I think this content fits in the documentation/src/main/asciidoc/stories/assembly_cache_configuration.adoc (chapter 3). I think that's where users will look for such information. The documentation/src/main/asciidoc/stories/assembly_cache_modes.adoc (chapter 1) summarizes use cases. These two assemblies are both in the same guide and aren't shared in other guides so I'd suggest removing it from chapter 1. WDYT @dfitzmau ?
> 
> HI @dvagnero . I wanted to put the notes in Chapter 1, so the notes display earlier in the guide to grab the user's attention.
> 
> I have removed the notes from chapter ; they now only display in Chapter 3. Please let me know if this is good.

Hi @dvagnero . Just a gentle reminder on the above?

dvagnero commented:
HI @dfitzmau seems like you removed the content from chapter 3 instead.

dvagnero commented on documentation/src/main/asciidoc/stories/assembly_cache_modes.adoc:
```diff
@@ -8,6 +8,25 @@
+
+Try to write succinct file names.
+====
+
```
```suggestion
```

dvagnero commented on documentation/src/main/asciidoc/stories/assembly_cache_modes.adoc:
```diff
@@ -8,6 +8,25 @@
+{brandname} places a restriction of less than or equal to 255 characters on a cache name or a cache template name.
+
+If you exceed this character limit, {brandname} throws an exception.
+
```
Why did you add these empty lines inside the admonitions?

dfitzmau replied:
From my experience on other projects, I do this to indicate paragraphs. Might I ask the impact of having blank lines? Do things break?

dvagnero commented on documentation/src/main/asciidoc/stories/assembly_cache_modes.adoc:
```diff
@@ -8,6 +8,25 @@
 
+[IMPORTANT]
+====
+{brandname} places a restriction of less than or equal to 255 characters on a cache name or a cache template name.
```
I think there is a bit of fluff
places restriction > restricts/limits
less than equal > maximum

dvagnero commented on documentation/src/main/asciidoc/stories/assembly_cache_modes.adoc:
```diff
@@ -8,6 +8,25 @@
+
+If you exceed this character limit, {brandname} throws an exception.
+
+Try to write succinct cache names and cache template names.
```
Try to write > write

dfitzmau commented on documentation/src/main/asciidoc/stories/assembly_cache_modes.adoc:
```diff
@@ -8,6 +8,25 @@
+
+[IMPORTANT]
+====
+A file system might set a limitation for the length of a file name, so ensure that a cache's name does not exceed this limitation.
```
Not possible to define, because this can vary depending on what an admin sets for a filename. We need to be vague here, because its outside our control.

dvagnero commented:
@dfitzmau I think I stated it above... Chapter 1 is about Infinispan use cases so adding this note there doesn't make much sense. Chapter 3 is about cache configuration and includes configuration examples with cache names https://infinispan.org/docs/stable/titles/configuring/configuring.html#cache-configuration_cache-configuration

dfitzmau commented:
> @dfitzmau I think I stated it above... Chapter 1 is about Infinispan use cases so adding this note there doesn't make much sense. Chapter 3 is about cache configuration and includes configuration examples with cache names https://infinispan.org/docs/stable/titles/configuring/configuring.html#cache-configuration_cache-configuration

OK. Updated and added to the section you specified. You OK to approve the PR?

dfitzmau commented:
> > @dfitzmau I think I stated it above... Chapter 1 is about Infinispan use cases so adding this note there doesn't make much sense. Chapter 3 is about cache configuration and includes configuration examples with cache names https://infinispan.org/docs/stable/titles/configuring/configuring.html#cache-configuration_cache-configuration
> 
> OK. Updated and added to the section you specified. You OK to approve the PR?

Thanks, @dvagnero . Can I merge this PR? Or do I need an additional reviewer? I understand that QE come into the equation post-merge of the PR.

dvagnero commented:
@dfitzmau there is the 14 release in progress. Check with the team please

dfitzmau commented:
> @dfitzmau there is the 14 release in progress. Check with the team please

Sorry I do not understand the question. When you say team as in docs team or the wider DG team?
